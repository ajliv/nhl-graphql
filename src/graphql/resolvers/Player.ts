import { GraphQLFieldResolver } from 'graphql';
import moment from 'moment';

import { Context } from '..';

export const Player: Record<
    string,
    GraphQLFieldResolver<Record<string, any>, Context>
> = {
    __resolveReference({ id }, ctx) {
        const { dataSources } = ctx as Context;
        return dataSources.nhl.getPerson(id);
    },

    age({ birthDate }, _, { now }) {
        if (birthDate) {
            return moment(now).diff(
                moment.utc(birthDate, 'YYYY-MM-DD'),
                'years'
            );
        }
    },

    birthDate({ birthDate }) {
        if (birthDate) {
            return moment.utc(birthDate, 'YYYY-MM-DD').toDate();
        }
    },

    currentTeam({ currentTeam }, _, { dataSources }) {
        return dataSources.nhl.getTeam(currentTeam?.id);
    },

    position({ primaryPosition }) {
        return primaryPosition?.abbreviation;
    },

    positionType({ primaryPosition }) {
        return primaryPosition?.type;
    },
};
