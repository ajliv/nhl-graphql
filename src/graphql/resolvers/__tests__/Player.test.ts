import { Player } from '../Player';
import { mockPerson, mockNHL } from '../../datasources/__tests__/fixtures';

const now = new Date(1569902400000); // 2019-10-01T00:00:00.000Z
const nhl = mockNHL();

const mockContext = {
    now,
    dataSources: {
        nhl,
    },
};

it('should return an age based on context `now`', async () => {
    const { age } = Player as any;
    const result = await age(mockPerson, {}, mockContext);
    expect(result).toMatchSnapshot();
});

it('should return a birthDate', async () => {
    const { birthDate } = Player as any;
    const result = await birthDate(mockPerson, {}, mockContext);
    expect(result).toMatchSnapshot();
});

it('should fetch a currentTeam', async () => {
    const { currentTeam } = Player as any;
    await currentTeam(mockPerson, {}, mockContext);
    expect(nhl.getTeam).toHaveBeenLastCalledWith(mockPerson.currentTeam.id);
});

it('should return a position', async () => {
    const { position } = Player as any;
    const result = await position(mockPerson, {}, mockContext);
    expect(result).toBe('LW');
});

it('should return a positionType', async () => {
    const { positionType } = Player as any;
    const result = await positionType(mockPerson, {}, mockContext);
    expect(result).toBe('Forward');
});
