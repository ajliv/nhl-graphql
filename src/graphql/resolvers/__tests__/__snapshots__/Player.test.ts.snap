// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return a birthDate 1`] = `1989-11-21T00:00:00.000Z`;

exports[`should return an age based on context \`now\` 1`] = `29`;
